Remove hard dependency on ADOL-C, Trilinos and SymEngine

jppelteret opened the pull request:
Fixes #35

Rombur commented:
Thanks!

jppelteret commented:
You're welcome. I hope that it works as-is, but since I've not tested it with any libraries excluded there might be a little bit of work to do before its working correctly.

Rombur commented on include/weak_forms/self_linearizing_forms.h:
```diff
@@ -819,6 +864,43 @@ namespace WeakForms
+      {
+        static_assert(false,
+                      "The SymEngine library is not enabled in deal.II.");
+        return void;
```
Why do you have this code? I have tried to run the tests with only Sacado installed and I hit this `static_assert` in a test. If I remove this code (and the other similar code), the test (weak_forms/energy_functor_01a) passes. I don't understand how this code is instantiated but is this code necessary? I would think that this is only there to get a better error message right? If I tried to call this code without `SymEngine`, I should get an error that the function does not exist, shouldn't I?
Here is the beginning of the error message. Like I said, it's not obvious to me why this code is instantiated:
```
20: ===============================   OUTPUT BEGIN  ===============================                                                                                                                                                          
20: weak_forms/energy_functor_01a.debug: BUILD failed. Output:                                                                                                                                                                               
20: [1/5] Building CXX object tests/weak_forms/CMakeFiles/weak_forms.energy_functor_01a.debug.dir/energy_functor_01a.cc.o                                                                                                                    
20: FAILED: tests/weak_forms/CMakeFiles/weak_forms.energy_functor_01a.debug.dir/energy_functor_01a.cc.o                                                                                                                                      
20: /usr/bin/c++ -DBOOST_NO_AUTO_PTR -DDEBUG -DSOURCE_DIR=\"/home/dealii-weak_forms/tests/weak_forms\" -DTESTING_ENVIRONMENT=light -I/home/dealii-weak_forms/include -I/home/dealii-weak_forms/build/configured_files -I/scratch/source/deali
i/tests -isystem /opt/dealii/include -isystem /opt/dealii/include/deal.II/bundled -isystem /opt/openmpi/include -isystem /opt/boost/include -isystem /opt/trilinos/include -isystem /opt/p4est/FAST/include -pedantic -fPIC -Wall -Wextra -Wm
issing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wswitch -Wsynth -Wwrite-strings -Wno-placement-new -Wno-literal-suffix -Wno-psabi -Wno-class-memaccess -fopenmp-simd -std=c++17 -pthread -Wno-parentheses -Wno-unused-loca
l-typedefs -O0 -ggdb -Wa,--compress-debug-sections -MD -MT tests/weak_forms/CMakeFiles/weak_forms.energy_functor_01a.debug.dir/energy_functor_01a.cc.o -MF tests/weak_forms/CMakeFiles/weak_forms.energy_functor_01a.debug.dir/energy_functor
_01a.cc.o.d -o tests/weak_forms/CMakeFiles/weak_forms.energy_functor_01a.debug.dir/energy_functor_01a.cc.o -c /home/dealii-weak_forms/tests/weak_forms/energy_functor_01a.cc                                                                 
20: In file included from /home/dealii-weak_forms/include/weak_forms/weak_forms.h:58,                                                                                                                                                        
20:                  from /home/dealii-weak_forms/tests/weak_forms/energy_functor_01a.cc:23:                                                                                                                                                 
20: /home/dealii-weak_forms/include/weak_forms/self_linearizing_forms.h: In member function 'auto dealiiWeakForms::WeakForms::SelfLinearization::EnergyFunctional<EnergyFunctor>::get_functor_first_derivative(const SymbolicOpField&, typena
me std::enable_if<dealiiWeakForms::WeakForms::is_sd_functor_op<T>::value>::type*) const':                                                                                                                                                    
20: /home/dealii-weak_forms/include/weak_forms/self_linearizing_forms.h:879:23: error: static assertion failed: The SymEngine library is not enabled in deal.II.                                                                             
20:   879 |         static_assert(false,                                                                                                                                                                                                     
20:       |                       ^~~~~                                                                                                                                                                                                      
20: /home/dealii-weak_forms/include/weak_forms/self_linearizing_forms.h:881:16: error: expected primary-expression before 'void'                                                                                                             
20:   881 |         return void;                                                                                                                                                                                                             
20:       |                ^~~~                                                                                                                                                                                                              
20: /home/dealii-weak_forms/include/weak_forms/self_linearizing_forms.h:881:15: error: expected ';' before 'void'                                                                                                                            
20:   881 |         return void;                                                                                                                                                                                                             
20:       |               ^~~~~                                                                                                                                                                                                              
20:       |               ;                                                                                                                                                                                                                  
20: /home/dealii-weak_forms/include/weak_forms/self_linearizing_forms.h:881:16: error: declaration does not declare anything [-fpermissive]                                                                                                  
20:   881 |         return void;                                                                                                                                                                                                             
20:       |                ^~~~ 
```

jppelteret commented:
> it's not obvious to me why this code is instantiated

Same for me. I'd hoped that having this static_assert would simply make life easier to understand what's going on in the case of some actually coding errors. Nevertheless, I've fixed the problem according to your feedback. I also split up one of the helper classes, which has maybe improved compilation time. Can you let me know if this now works for you? It still seems to be working with all of the optional libraries enabled.

Rombur commented:
> Can you let me know if this now works for you?

Yes, I will try the PR with Sacado only. I ran the tests overnight after commenting the code with the static_assert and the only tests that failed at the time were
```
        110 - weak_forms/step-transient_curl_curl-variant_01a.mpirun=1.debug (Failed)
        111 - weak_forms/step-transient_curl_curl-variant_01a.mpirun=3.debug (Failed)
        112 - weak_forms/step-transient_curl_curl-variant_01b.mpirun=1.debug (Failed)
        113 - weak_forms/step-transient_curl_curl-variant_01b.mpirun=3.debug (Failed)
        114 - weak_forms/step-transient_curl_curl-variant_02a.mpirun=1.debug (Failed)
        115 - weak_forms/step-transient_curl_curl-variant_02a.mpirun=3.debug (Failed)
        116 - weak_forms/step-transient_curl_curl-variant_02b.mpirun=1.debug (Failed)
        117 - weak_forms/step-transient_curl_curl-variant_02b.mpirun=3.debug (Failed)
```
It looks like the errors are due to the position of the points in the reference cell. Same problem as https://github.com/jppelteret/dealii-weak_forms/pull/43#issuecomment-1092358481

jppelteret commented:
> It looks like the errors are due to the position of the points in the reference cell. Same problem as https://github.com/jppelteret/dealii-weak_forms/pull/43#issuecomment-1092358481

Makes sense -- I'll get on to this next. I wasn't feeling sufficiently enthused to tackle this last weekend. I want to do some larger changes to that test, but I'll just put the test in a reliably-working state first before reformulating it.

jppelteret commented:
@Rombur I followed through on your suggestion to remove the `Point` from the outputted data in the `step-transient-curl-curl` test, and I added that change to this branch. It would be interesting to know if that now makes the output completely on machines other than my Mac.

Rombur commented:
Now everything passes but 
```
step-transient_curl_curl-variant_02b.mpirun=3
```
It looks that this test has two outputs https://github.com/jppelteret/dealii-weak_forms/blob/main/tests/weak_forms/step-transient_curl_curl-variant_02b.with_trilinos%3Don.mpirun%3D3.output and https://github.com/jppelteret/dealii-weak_forms/blob/main/tests/weak_forms/step-transient_curl_curl-variant_02b.with_trilinos%3Don.mpirun%3D3.output.2 What I see is closer to `output.2` but it is different. The error seems to be `J_eddy (axial)`
